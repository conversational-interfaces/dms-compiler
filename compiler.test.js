const compile = require('./compiler');

describe('Compiler', () => {
  test('if', () => {
    expect(
      compile(`
        if X == 0 {
          act First
        }
        else {
          act Second
        }
    `)
    ).toMatchSnapshot();

    expect(
      compile(`
        if X == 0 {
          act First
        }
        else if X == 1 {
          act Second
        }
        else {
          act DoNothing
        }
    `)
    ).toMatchSnapshot();
  });

  test('fork', () => {
    expect(
      compile(`
      fork {
        X == 0 {
          act First
        }
        X == 1 {
          act Second
        }
        _ {
          act Third
        }
        _ {
          act Fourth
        }
      }
    `)
    ).toMatchSnapshot();

    expect(
      compile(`
      #{depth: 2}
      fork {
        X == 0 {
          act First
        }
        X == 1 {
          act Second
        }
      }
    `)
    ).toMatchSnapshot();
  });

  test('input', () => {
    expect(
      compile(`
      input -> result {
        result == 0 {
          act First
        }
        result == 1 {
          act Second
        }
        _ {
          act Third
        }
      }
    `)
    ).toMatchSnapshot();
  });

  test('await', () => {
    expect(compile('await X == 5 { act Hello }')).toMatchSnapshot();
  });

  test('once', () => {
    expect(compile('once { act Hello }')).toMatchSnapshot();
  });

  test('act', () => {
    expect(compile('act One \n act Two \n act Three')).toMatchSnapshot();
  });

  test('set', () => {
    expect(compile('X = value')).toMatchSnapshot();
    expect(compile('X, Y, Z = value')).toMatchSnapshot();
    expect(compile('a[0]["a"] = 1')).toMatchSnapshot();
    expect(compile('d[a] = 1')).toMatchSnapshot();
  });

  test('pop', () => {
    expect(compile('X = 50 \n pop X')).toMatchSnapshot();
  });

  test('run', () => {
    expect(compile('run "Questions" (foo, bar)')).toMatchSnapshot();
    expect(
      compile(`
      run "Questions" (foo, bar) -> result {
        _ { act result }
      }
    `)
    ).toMatchSnapshot();
    expect(compile('run "MyComponent"')).toMatchSnapshot();
    expect(compile('run "MyComponent" ()')).toMatchSnapshot();
    expect(compile('run "MyComponent" () -> result')).toMatchSnapshot();
    expect(compile(`
      run "MyComponent" () -> _result {
        _ {
            result = _result
        }
      }
    `)).toMatchSnapshot();
  });

  test('use', () => {
    expect(compile('use "Questions" import foo, bar')).toMatchSnapshot();
  });

  test('def', () => {
    expect(
      compile(`
        def Pow(x) {
          pop x * x
        }

        act Pow(5)
    `)
    ).toMatchSnapshot();
  });

  test('comments', () => {
    expect(
      compile(`
        // This is a comment
        if X == 0 {
          // This is another comment
          act First
        }
        else {
          act Second
        }
    `)
    ).toMatchSnapshot();

    expect(
      compile(`
        if X == 0 {
          act 1 +
            // This is a comment
            (4 * 10) /
            // This is another comment
            pow(5)
        }
    `)
    ).toMatchSnapshot();
  });

  test('expressions', () => {
    expect(compile('act X + 3 * 5')).toMatchSnapshot();
    expect(compile('act [1, 2, 3, 4]')).toMatchSnapshot();
    expect(compile('act "string" ++ "string2"')).toMatchSnapshot();
    expect(compile('act 1 < 2 && 3 < 4 || 5 >= 6')).toMatchSnapshot();
    expect(compile('act 1 in [1, 2, 3, 4]')).toMatchSnapshot();
    expect(compile('act funcall({ foo: 1, bar: 2})')).toMatchSnapshot();
    expect(compile('act !TruthyValue')).toMatchSnapshot();
    expect(compile('act exists(TruthyValue)')).toMatchSnapshot();
    expect(compile('X = null')).toMatchSnapshot();

    // Member lookup
    expect(compile('act X[0]')).toMatchSnapshot();
    expect(compile('act X["str"]')).toMatchSnapshot();
  });

  test('init dictionary', () => {
    expect(compile('a = {1: 12, "-1": -1, "a": 11, a: 11}')).toMatchSnapshot();
  });

  test('Planet Trivia example', () => {
    let text = `
      if !exists(Init) {
        Init = true
        WrongCounter = 0
        IsHint1Given = false
        IsHint2Given = false
        IsAnswerGiven = false
        NumHintsGiven = 0
        Question = "What's the biggest planet?"
        CA = "planet.jupiter"
        CAResponse = "Exactly!"
        WA1 = "planet"
        WA1Response = "Nope. That's not the biggest"
        WA2 = "nonplanet"
        WA2Response = "That's not a planet"
        Hint1 = "It has a big red spot"
        Hint2 = "It's name begins with the letter J"
        Answer = "The biggest planet is Jupiter"
        HintAnnouncement = "Here's a hint"
      }
      else {
        #{depth: 3}
        fork {
          WrongCounter <= 3 && !IsAnswerGiven  {
            act Question
            input -> result {
              result == CA {
                act CAResponse
                IsAnswerGiven = true
              }
              result == WA1 {
                act WA1Response
                WrongCounter = WrongCounter + 1
              }
              result == WA2 {
                act WA2Response
                WrongCounter = WrongCounter + 1
              }
              _ {
                WrongCounter = WrongCounter + 1
              }
            }
          }
          !IsHint1Given && WrongCounter == 1 {
            act HintAnnouncement
            act Hint1
            IsHint1Given = true
          }
          !IsHint2Given && WrongCounter == 2 {
            act HintAnnouncement
            act Hint2
            IsHint2Given = true
          }
          !IsAnswerGiven && WrongCounter == 3 {
            act Answer
            IsAnswerGiven = true
          }
        }
      }
    `;

    expect(compile(text)).toMatchSnapshot();
  });

  test('chaining member lookups and function calls', () => {
    expect(compile('x = d["a"]["b"]["c"]')).toMatchSnapshot();
    expect(compile('x = d(a)(b)(c)')).toMatchSnapshot();
    expect(compile('x = d(a)["b"](c)["d"]')).toMatchSnapshot();
  });

  test('builtin pref function', () => {
    let output = compile(`
        utility = [
            pref({meeting_scheduled: true}),
            pref({meeting_scheduled: true, schedule_conflict: false}),
            pref({meeting_scheduled: true, schedule_conflict: false}, ["schedule_conflict"]),
            pref({polite: true, refuse_counter: 0}, ["refuse_counter"]),
            pref({int: 10, int2: 3})
        ]
    `);
    expect(output).toMatchSnapshot();
  });

  test('hop', () => {
    let output = compile(`
      hop 3
    `);
    expect(output).toMatchSnapshot();
  });
});
