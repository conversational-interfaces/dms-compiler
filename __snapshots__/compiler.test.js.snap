// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler Planet Trivia example 1`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@set": "\`Init\`",
              "val": true,
            },
            Object {
              "@set": "\`WrongCounter\`",
              "val": 0,
            },
            Object {
              "@set": "\`IsHint1Given\`",
              "val": false,
            },
            Object {
              "@set": "\`IsHint2Given\`",
              "val": false,
            },
            Object {
              "@set": "\`IsAnswerGiven\`",
              "val": false,
            },
            Object {
              "@set": "\`NumHintsGiven\`",
              "val": 0,
            },
            Object {
              "@set": "\`Question\`",
              "val": "\`What's the biggest planet?\`",
            },
            Object {
              "@set": "\`CA\`",
              "val": "\`planet.jupiter\`",
            },
            Object {
              "@set": "\`CAResponse\`",
              "val": "\`Exactly!\`",
            },
            Object {
              "@set": "\`WA1\`",
              "val": "\`planet\`",
            },
            Object {
              "@set": "\`WA1Response\`",
              "val": "\`Nope. That's not the biggest\`",
            },
            Object {
              "@set": "\`WA2\`",
              "val": "\`nonplanet\`",
            },
            Object {
              "@set": "\`WA2Response\`",
              "val": "\`That's not a planet\`",
            },
            Object {
              "@set": "\`Hint1\`",
              "val": "\`It has a big red spot\`",
            },
            Object {
              "@set": "\`Hint2\`",
              "val": "\`It's name begins with the letter J\`",
            },
            Object {
              "@set": "\`Answer\`",
              "val": "\`The biggest planet is Jupiter\`",
            },
            Object {
              "@set": "\`HintAnnouncement\`",
              "val": "\`Here's a hint\`",
            },
          ],
          "if": Array [
            "!",
            Array [
              "?",
              "Init",
            ],
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@fork": Array [
                Object {
                  "@do": Array [
                    Object {
                      "@act": "Question",
                    },
                    Object {
                      "@fork": Array [
                        Object {
                          "@do": Array [
                            Object {
                              "@act": "CAResponse",
                            },
                            Object {
                              "@set": "\`IsAnswerGiven\`",
                              "val": true,
                            },
                          ],
                          "if": Array [
                            "==",
                            "_nlu",
                            "CA",
                          ],
                        },
                        Object {
                          "@do": Array [
                            Object {
                              "@act": "WA1Response",
                            },
                            Object {
                              "@set": "\`WrongCounter\`",
                              "val": Array [
                                "+",
                                "WrongCounter",
                                1,
                              ],
                            },
                          ],
                          "if": Array [
                            "==",
                            "_nlu",
                            "WA1",
                          ],
                        },
                        Object {
                          "@do": Array [
                            Object {
                              "@act": "WA2Response",
                            },
                            Object {
                              "@set": "\`WrongCounter\`",
                              "val": Array [
                                "+",
                                "WrongCounter",
                                1,
                              ],
                            },
                          ],
                          "if": Array [
                            "==",
                            "_nlu",
                            "WA2",
                          ],
                        },
                        Object {
                          "@do": Array [
                            Object {
                              "@set": "\`WrongCounter\`",
                              "val": Array [
                                "+",
                                "WrongCounter",
                                1,
                              ],
                            },
                          ],
                        },
                      ],
                      "await": Array [
                        "input",
                      ],
                    },
                  ],
                  "if": Array [
                    "&&",
                    Array [
                      "<=",
                      "WrongCounter",
                      3,
                    ],
                    Array [
                      "!",
                      "IsAnswerGiven",
                    ],
                  ],
                },
                Object {
                  "@do": Array [
                    Object {
                      "@act": "HintAnnouncement",
                    },
                    Object {
                      "@act": "Hint1",
                    },
                    Object {
                      "@set": "\`IsHint1Given\`",
                      "val": true,
                    },
                  ],
                  "if": Array [
                    "&&",
                    Array [
                      "!",
                      "IsHint1Given",
                    ],
                    Array [
                      "==",
                      "WrongCounter",
                      1,
                    ],
                  ],
                },
                Object {
                  "@do": Array [
                    Object {
                      "@act": "HintAnnouncement",
                    },
                    Object {
                      "@act": "Hint2",
                    },
                    Object {
                      "@set": "\`IsHint2Given\`",
                      "val": true,
                    },
                  ],
                  "if": Array [
                    "&&",
                    Array [
                      "!",
                      "IsHint2Given",
                    ],
                    Array [
                      "==",
                      "WrongCounter",
                      2,
                    ],
                  ],
                },
                Object {
                  "@do": Array [
                    Object {
                      "@act": "Answer",
                    },
                    Object {
                      "@set": "\`IsAnswerGiven\`",
                      "val": true,
                    },
                  ],
                  "if": Array [
                    "&&",
                    Array [
                      "!",
                      "IsAnswerGiven",
                    ],
                    Array [
                      "==",
                      "WrongCounter",
                      3,
                    ],
                  ],
                },
              ],
              "scheme": Object {
                "\`depth\`": 3,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler act 1`] = `
Object {
  "@do": Array [
    Object {
      "@act": "One",
    },
    Object {
      "@act": "Two",
    },
    Object {
      "@act": "Three",
    },
  ],
}
`;

exports[`Compiler await 1`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@act": "Hello",
        },
      ],
      "await": Array [
        "==",
        "X",
        5,
      ],
    },
  ],
}
`;

exports[`Compiler builtin pref function 1`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`utility\`",
      "val": Array [
        "",
        Array [
          "",
          Object {
            "\`meeting_scheduled\`": true,
          },
          Object {
            "\`meeting_scheduled\`": false,
          },
        ],
        Array [
          "",
          Object {
            "\`meeting_scheduled\`": true,
            "\`schedule_conflict\`": false,
          },
          Object {
            "\`meeting_scheduled\`": false,
            "\`schedule_conflict\`": true,
          },
        ],
        Array [
          "",
          Object {
            "\`meeting_scheduled\`": true,
            "\`schedule_conflict\`": false,
          },
          Object {
            "\`meeting_scheduled\`": true,
            "\`schedule_conflict\`": true,
          },
        ],
        Array [
          "",
          Object {
            "\`polite\`": true,
            "\`refuse_counter\`": 0,
          },
          Object {
            "\`polite\`": true,
            "\`refuse_counter\`": 1,
          },
        ],
        Array [
          "",
          Object {
            "\`polite\`": true,
            "\`refuse_counter\`": 0,
          },
          Object {
            "\`polite\`": true,
            "\`refuse_counter\`": -1,
          },
        ],
        Array [
          "",
          Object {
            "\`int2\`": 3,
            "\`int\`": 10,
          },
          Object {
            "\`int2\`": 12,
            "\`int\`": 40,
          },
        ],
        Array [
          "",
          Object {
            "\`int2\`": 3,
            "\`int\`": 10,
          },
          Object {
            "\`int2\`": -6,
            "\`int\`": -20,
          },
        ],
      ],
    },
  ],
}
`;

exports[`Compiler chaining member lookups and function calls 1`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`x\`",
      "val": Array [
        "get",
        "\`c\`",
        Array [
          "get",
          "\`b\`",
          Array [
            "get",
            "\`a\`",
            "d",
          ],
        ],
      ],
    },
  ],
}
`;

exports[`Compiler chaining member lookups and function calls 2`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`x\`",
      "val": Array [
        Array [
          Array [
            "d",
            "a",
          ],
          "b",
        ],
        "c",
      ],
    },
  ],
}
`;

exports[`Compiler chaining member lookups and function calls 3`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`x\`",
      "val": Array [
        "get",
        "\`d\`",
        Array [
          Array [
            "get",
            "\`b\`",
            Array [
              "d",
              "a",
            ],
          ],
          "c",
        ],
      ],
    },
  ],
}
`;

exports[`Compiler comments 1`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@act": "First",
            },
          ],
          "if": Array [
            "==",
            "X",
            0,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Second",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler comments 2`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "+",
        1,
        Array [
          "/",
          Array [
            "*",
            4,
            10,
          ],
          Array [
            "pow",
            5,
          ],
        ],
      ],
    },
  ],
  "if": Array [
    "==",
    "X",
    0,
  ],
}
`;

exports[`Compiler def 1`] = `
Object {
  "@do": Array [
    Object {
      "@def": Array [
        "",
        "\`Pow\`",
        "\`x\`",
      ],
      "val": Object {
        "@do": Array [
          Object {
            "@pop": Array [
              "*",
              "x",
              "x",
            ],
          },
        ],
      },
    },
    Object {
      "@act": Array [
        "Pow",
        5,
      ],
    },
  ],
}
`;

exports[`Compiler expressions 1`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "+",
        "X",
        Array [
          "*",
          3,
          5,
        ],
      ],
    },
  ],
}
`;

exports[`Compiler expressions 2`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "",
        1,
        2,
        3,
        4,
      ],
    },
  ],
}
`;

exports[`Compiler expressions 3`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "++",
        "\`string\`",
        "\`string2\`",
      ],
    },
  ],
}
`;

exports[`Compiler expressions 4`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "&&",
        Array [
          "<",
          1,
          2,
        ],
        Array [
          "||",
          Array [
            "<",
            3,
            4,
          ],
          Array [
            ">=",
            5,
            6,
          ],
        ],
      ],
    },
  ],
}
`;

exports[`Compiler expressions 5`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "in",
        1,
        Array [
          "",
          1,
          2,
          3,
          4,
        ],
      ],
    },
  ],
}
`;

exports[`Compiler expressions 6`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "funcall",
        Object {
          "\`bar\`": 2,
          "\`foo\`": 1,
        },
      ],
    },
  ],
}
`;

exports[`Compiler expressions 7`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "!",
        "TruthyValue",
      ],
    },
  ],
}
`;

exports[`Compiler expressions 8`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "?",
        "TruthyValue",
      ],
    },
  ],
}
`;

exports[`Compiler expressions 9`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`X\`",
      "val": null,
    },
  ],
}
`;

exports[`Compiler expressions 10`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "get",
        0,
        "X",
      ],
    },
  ],
}
`;

exports[`Compiler expressions 11`] = `
Object {
  "@do": Array [
    Object {
      "@act": Array [
        "get",
        "\`str\`",
        "X",
      ],
    },
  ],
}
`;

exports[`Compiler fork 1`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@act": "First",
            },
          ],
          "if": Array [
            "==",
            "X",
            0,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Second",
            },
          ],
          "if": Array [
            "==",
            "X",
            1,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Third",
            },
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Fourth",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler fork 2`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@act": "First",
            },
          ],
          "if": Array [
            "==",
            "X",
            0,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Second",
            },
          ],
          "if": Array [
            "==",
            "X",
            1,
          ],
        },
      ],
      "scheme": Object {
        "\`depth\`": 2,
      },
    },
  ],
}
`;

exports[`Compiler hop 1`] = `
Object {
  "@do": Array [
    Object {
      "@hop": 3,
    },
  ],
}
`;

exports[`Compiler if 1`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@act": "First",
            },
          ],
          "if": Array [
            "==",
            "X",
            0,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Second",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler if 2`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@act": "First",
            },
          ],
          "if": Array [
            "==",
            "X",
            0,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Second",
            },
          ],
          "if": Array [
            "==",
            "X",
            1,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "DoNothing",
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler init dictionary 1`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`a\`",
      "val": Object {
        "\`-1\`": -1,
        "\`1\`": 12,
        "\`a\`": 11,
      },
    },
  ],
}
`;

exports[`Compiler input 1`] = `
Object {
  "@do": Array [
    Object {
      "@fork": Array [
        Object {
          "@do": Array [
            Object {
              "@act": "First",
            },
          ],
          "if": Array [
            "==",
            "_nlu",
            0,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Second",
            },
          ],
          "if": Array [
            "==",
            "_nlu",
            1,
          ],
        },
        Object {
          "@do": Array [
            Object {
              "@act": "Third",
            },
          ],
        },
      ],
      "await": Array [
        "input",
      ],
    },
  ],
}
`;

exports[`Compiler once 1`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@act": "Hello",
        },
      ],
      "once": true,
    },
  ],
}
`;

exports[`Compiler pop 1`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`X\`",
      "val": 50,
    },
    Object {
      "@pop": "X",
    },
  ],
}
`;

exports[`Compiler run 1`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@run": "\`Questions\`",
          "args": Array [
            "",
            "foo",
            "bar",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler run 2`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@run": "\`Questions\`",
          "args": Array [
            "",
            "foo",
            "bar",
          ],
        },
        Object {
          "@fork": Array [
            Object {
              "@do": Array [
                Object {
                  "@act": "_return",
                },
              ],
            },
          ],
          "await": Array [
            "return",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler run 3`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@run": "\`MyComponent\`",
          "args": Array [
            "",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler run 4`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@run": "\`MyComponent\`",
          "args": Array [
            "",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler run 5`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@run": "\`MyComponent\`",
          "args": Array [
            "",
          ],
        },
        Object {
          "@fork": Array [
            Object {
              "@do": Array [
                Object {
                  "@set": "\`result\`",
                  "val": "_return",
                },
              ],
            },
          ],
          "await": Array [
            "return",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler run 6`] = `
Object {
  "@do": Array [
    Object {
      "@do": Array [
        Object {
          "@run": "\`MyComponent\`",
          "args": Array [
            "",
          ],
        },
        Object {
          "@fork": Array [
            Object {
              "@do": Array [
                Object {
                  "@set": "\`result\`",
                  "val": "_return",
                },
              ],
            },
          ],
          "await": Array [
            "return",
          ],
        },
      ],
    },
  ],
}
`;

exports[`Compiler set 1`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`X\`",
      "val": "value",
    },
  ],
}
`;

exports[`Compiler set 2`] = `
Object {
  "@do": Array [
    Object {
      "@set": Array [
        "",
        "\`X\`",
        "\`Y\`",
        "\`Z\`",
      ],
      "val": "value",
    },
  ],
}
`;

exports[`Compiler set 3`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`a\`",
      "val": Array [
        "\`edit\`",
        "a",
        1,
        0,
        "\`a\`",
      ],
    },
  ],
}
`;

exports[`Compiler set 4`] = `
Object {
  "@do": Array [
    Object {
      "@set": "\`d\`",
      "val": Array [
        "\`edit\`",
        "d",
        1,
        "a",
      ],
    },
  ],
}
`;

exports[`Compiler use 1`] = `
Object {
  "@do": Array [
    Object {
      "@use": "\`Questions\`",
      "import": Array [
        "",
        "\`foo\`",
        "\`bar\`",
      ],
    },
  ],
}
`;
